velox/functions/lib/DateTimeFormatter.h

```diff
@@ -0,0 +1,371 @@
+
+  // Year of era (>=0), e.g: 1996
+  WEEK_YEAR = 3,
+
+  // Week of weekyear, e.g: 27
+  WEEK_OF_WEEK_YEAR = 4,
+
+  // Day of week, e.g: 2
+  DAY_OF_WEEK = 5,
+
+  // Day of week, e.g: "Tuesday" or "Tue", depending on number of times provided
+  // formatter character repeats
+  DAY_OF_WEEK_TEXT = 6,
+
+  // Year, e.g: 1996
+  YEAR = 7,
+
+  // Day of year, e.g: 189
+  DAY_OF_YEAR = 8,
+
+  // Month of year, e.g: "July", "Jul", 07, or 7 depending on number of times
+  // provided formatter character repeats
+  MONTH_OF_YEAR = 9,
+
+  // Day of month, e.g: 10, 01, 001, with/without padding 0s depending on number
+  // of times provided formatter character repeats
+  DAY_OF_MONTH = 10,
+
+  // Halfday of day, e.g: "PM"
+  HALFDAY_OF_DAY = 11,
+
+  // Hour of halfday (0~11)
+  HOUR_OF_HALFDAY = 12,
+
+  // Clockhour of halfday (1~12)
+  CLOCK_HOUR_OF_HALFDAY = 13,
+
+  // Hour of day (0~23)
+  HOUR_OF_DAY = 14,
+
+  // Clockhour of day (1~24)
+  CLOCK_HOUR_OF_DAY = 15,
+
+  // Minute of hour, e.g: 30
+  MINUTE_OF_HOUR = 16,
+
+  // Second of minute, e.g: 55
+  SECOND_OF_MINUTE = 17,
+
+  // Fraction of second, e.g: 978
+  FRACTION_OF_SECOND = 18,
+
+  // Timezone, e.g: "Pacific Standard Time" or "PST"
+  TIMEZONE = 19,
+
+  // Timezone offset/id, e.g: "-0800", "-08:00" or "America/Los_Angeles"
+  TIMEZONE_OFFSET_ID = 20,
+
+  // A literal % character
+  LITERAL_PERCENT = 21
+};
+
+struct FormatPattern {
+  DateTimeFormatSpecifier specifier;
+  size_t minRepresentDigits;
+};
+
+struct DateTimeToken {
+  enum class Type { kPattern, kLiteral } type;
+  union {
+    FormatPattern pattern;
+    std::string_view literal;
+  };
+
+  explicit DateTimeToken(const FormatPattern& pattern)
+      : type(Type::kPattern), pattern(pattern) {}
+
+  explicit DateTimeToken(const std::string_view& literal)
+      : type(Type::kLiteral), literal(literal) {}
+
+  bool operator==(const DateTimeToken& right) const {
+    if (type == right.type) {
+      if (type == Type::kLiteral) {
+        return literal == right.literal;
+      } else {
+        return pattern.specifier == right.pattern.specifier &&
+            pattern.minRepresentDigits == right.pattern.minRepresentDigits;
+      }
+    }
+    return false;
+  }
+};
+
+struct DateTimeResult {
+  Timestamp timestamp;
+  int64_t timezoneId{-1};
+};
+
+class DateTimeFormatter {
+ public:
+  explicit DateTimeFormatter(
+      std::unique_ptr<char[]>&& literalBuf,
+      size_t bufSize,
+      std::vector<DateTimeToken>&& tokens)
+      : literalBuf_(std::move(literalBuf)),
+        bufSize_(bufSize),
+        tokens_(std::move(tokens)) {}
+
+  const std::unique_ptr<char[]>& literalBuf() const {
+    return literalBuf_;
+  }
+
+  size_t bufSize() const {
+    return bufSize_;
+  }
+
+  const std::vector<DateTimeToken>& tokens() const {
+    return tokens_;
+  }
+
+ private:
+  std::unique_ptr<char[]> literalBuf_;
+  size_t bufSize_;
+  std::vector<DateTimeToken> tokens_;
+};
+
+class DateTimeFormatterBuilder {
```

Comment:
if the idea now is to expose the API through this builder class, shouldn't the enum be hidden inside the .cpp file?

Reply:
The API will be through DateTimeFormatter class.

Reply:
So, the builder would not even be required in the code?
Then it makes sense to move that builder to a separate h/cpp file and use it inly in tests and DateTimeFormatter.cpp